src/SFA.DAS.EmployerIncentives.Api/Controllers/WithdrawalCommandController.cs

```diff
@@ -11,15 +14,29 @@ namespace SFA.DAS.EmployerIncentives.Api.Controllers
     [ApiController]
     public class WithdrawalCommandController : ApiCommandControllerBase
     {
-        public WithdrawalCommandController(ICommandDispatcher commandDispatcher) : base(commandDispatcher)
+        private readonly IApprenticeshipIncentiveQueryRepository _queryRepository;
+
+        public WithdrawalCommandController(
+            ICommandDispatcher commandDispatcher,
+            IApprenticeshipIncentiveQueryRepository queryRepository) : base(commandDispatcher)
         {
+            _queryRepository = queryRepository;
         }
 
         [HttpPost("/withdrawals")]
         [ProducesResponseType((int)HttpStatusCode.Accepted)]
         [ProducesResponseType((int)HttpStatusCode.BadRequest)]
         public async Task<IActionResult> WithdrawalIncentiveApplication([FromBody] WithdrawApplicationRequest request)
         {
+            var incentiveModel = await _queryRepository.Get((a => 
+                a.AccountLegalEntityId == request.AccountLegalEntityId &&
+                a.ULN == request.ULN), includePayments:true);
+            
+            if(incentiveModel != null && incentiveModel.Payments.Any())
+            {
+                return BadRequest(new { Error = "Cannot withdraw an application that has been submitted and has received payments" });
+            }
+
```

Comment:
It doesn't feel right to be calling the repository directly from the controller. Would it be better that the handler performs this check and bubbles an exception up to the controller, which could then be returned as a BadRequest response?

Reply:
It's in a different domain so the handler isn't the correct place.  If the repository is considered as a read only model then I would say it's OK to be called since the client can't change the data.  It's quite common to have a read only repository return value objects as part of CQRS.  e.g. https://stackoverflow.com/questions/47879119/ddd-read-only-repositories-returning-value-objects

Reply:
Makes sense.

Reply:
can't a QueryHandler be used in a CommandController?